### PR TITLE
Make GetEncodedUrl consistent

### LINKED ISSUE #1603: Allow Multi Part Tenant URLs
Currently  if you try to use a slash in a tenant URL it will give an error disallowing you to use the url 
ie http://www.testsite.local:8080/tenants/MyTenantSite

I f you try to force it in by using the domain name and prefix it completely breaks,  and in some cases put the port number in the wrong place if using a port number in the domain prefix 

 ie entering localhost:1058/artists/ as the hostname results in urls being misformatted to http://localhost:1058/artists:1058/krista-herzog

if try using the expected url it redirects to the root site 
ie http://localhost:1058/artists/krista-herzog  return the content for  http://localhost:1058/

image

### PR DESCRIPTION
**Pull request by infofromca**
Fixes #1603

### CODE REVIEW

**Comment by jtkech** [comment]
@infofromca 

Sorry, i already did some `HostString` tweaks recently, but forgot to do the same in the `ApiController`

You remembered me this missing part, thanks

So i will do it and you will have some conflicts to solve, sorry

But still good to suggest a shared service for this

File: src/OrchardCore/OrchardCore.Setup.Core/UrlService.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                hostString = new HostString(tenantUrlHost);
+            }
+
+            var pathString = httpContext.Features.Get<ShellContextFeature>().OriginalPathBase;
+            if (!String.IsNullOrEmpty(shellSettings.RequestUrlPrefix))
+            {
+                pathString = pathString.Add('/' + shellSettings.RequestUrlPrefix);
+            }
+
+            baseUrl = $"{httpContext.Request.Scheme}://{hostString.ToUriComponent() + pathString.ToUriComponent()}";
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Setup.Core/UrlService.cs:39]: For info here no need to have `ToUriComponent()` (that url encodes the related uri component), here the `ToString()` is implicitly called, and it calls `ToUriComponent()`.

Then i like to have in one line at the end what aspnetcore also uses, but not always ;)

            $"{Request.Scheme}://{hostString + pathString + queryString}";

Then, to build a non encoded url for display only, you have to use the `.Value` property of each component, which is the original unencoded (unescaped) string.

Unless for a `QueryString` component whose ctor needs an already escaped string ;) So not well adapted for display only

        /// Initialize the query string with a given value. This value must be in escaped and delimited format with
        /// a leading '?' character.
        /// </summary>
        /// <param name="value">The query string to be assigned to the Value property.</param>
        public QueryString(string? value)
        {
            if (!string.IsNullOrEmpty(value) && value[0] != '?')
            {
                throw new ArgumentException("The leading '?' must be included for a non-empty query.", nameof(value));
            }
            Value = value;
        }

**Comment by jtkech** [comment]
@infofromca 

Sorry for the conflicts

thanks to have pointed out the missing updates in the ApiController

**Comment by agriffard** [comment]
#1603 is already closed.

Is this PR still relevant?

**Comment by Piedone** [comment]
Closing due to no reply from @infofromca.

**Pull request closed without merging**